UW-586: Build driver for WaveWatchIII (#496)
ufs-community/uwtools: docs/sections/user_guide/yaml/components/ww3.rst

.. _ww3_yaml:

ww3
===

> Structured YAML when including WaveWatchIII as part of a compiled coupled executable is validated by JSON Schema and requires the ``ww3:`` block, described below.

Reviewer: ```suggestion
Structured YAML to configure WaveWatchIII as part of a compiled coupled executable is validated by JSON Schema and requires the ``ww3:`` block, described below.
```

Here is a prototype UW YAML ``ww3:`` block, explained in detail below:

.. highlight:: yaml
.. literalinclude:: ../../../../shared/ww3.yaml

UW YAML for the ``ww3:`` Block
------------------------------

> namelists:

Reviewer 2: ```suggestion
namelist:
```

In the shared ww3.yaml the block is called `namelist:`

Author: I knew I was gonna miss one when I changed that, thanks!

^^^^^^^^^^

> The path to the namelist(s) used by the WaveWatchIII configuration. Namelists will be symlinked to the run directory.

Reviewer: Did we decide to render templates into the run directory instead of linking files?

Author: Ah, I didn't change this after implementing it to look more like the `fv3` driver on suggestion, which does currently render the templates rather than linking

.. _ww3_yaml:

ww3
===

Structured YAML to configure WaveWatchIII as part of a compiled coupled executable is validated by JSON Schema and requires the ``ww3:`` block, described below.

Here is a prototype UW YAML ``ww3:`` block, explained in detail below:

.. highlight:: yaml
.. literalinclude:: ../../../../shared/ww3.yaml

UW YAML for the ``ww3:`` Block
------------------------------

namelist:
^^^^^^^^^

The WaveWatchIII namelist file can be provisioned either by:
1. Providing a path to a complete, ready-to-use namelist file as the value of the `base_file:` key and omitting the `update_values:` key, or
2. Providing a path to a namelist file containing Jinja2 expressions as the value of the `base_file:` key and a mapping from variable names to values used to render those Jinja2 expressions as the value of the `update_values:` key.

> 

Reviewer 3: Given the ongoing discussion below about the names of the subkeys in this section, I think it might be helpful here to provide a warning note about the behavior in this section not conforming to the same behavior as the other namelist sections.

.. _ww3_yaml:

ww3
===

Structured YAML to configure WaveWatchIII as part of a compiled coupled executable is validated by JSON Schema and requires the ``ww3:`` block, described below.

Here is a prototype UW YAML ``ww3:`` block, explained in detail below:

.. highlight:: yaml
.. literalinclude:: ../../../../shared/ww3.yaml

UW YAML for the ``ww3:`` Block
------------------------------

namelist:
^^^^^^^^^

> .. important:: Fortran namelists for WaveWatchIII are used here as templates to be realized. Input namelists will not have their configurations validated, only the associated YAML. Other files are not currently supported.:

Author: I wanted to leave something to note that we are specifically not supporting the inp.in file, even though it is accepted by WW3, but maybe its sufficiently implied here?

> .. important:: Fortran namelists for WaveWatchIII are used here as templates to be realized. Input namelists will not have their configurations validated, only the associated YAML. Other files are not currently supported.:

Reviewer: That could be good advice, if made explicit. In the `template_file` section below, we specifically mention the `ww3_shel.nml` file, so maybe that would be a good place for an additional sentence like
```
Note that the non-namelist ``ww3_shel.inp`` file may not be used as the basis for the input template file.
```
?

> .. important:: Fortran namelists for WaveWatchIII are used here as templates to be realized. Input namelists will not have their configurations validated, only the associated YAML. Other files are not currently supported.:

Author: OK, yeah - no need to be coy about it, just be decisive

  **template_file:**

>   The path to the base ``ww3_shel.nml`` file that comes from the WW3 build.

Reviewer: ```suggestion
  The path to the input template file containing Jinja2 expressions, based on the ``ww3_shel.nml`` file from the WaveWatchIII build.
```
True?

Author: Currently this is pointing directly to the namelist `ww3_shel.nml`.

Reviewer: How is it currently pointing to anything? The user has to set `template_file:` in their UW YAML to specify the input template file they want us to use, right? We'd just be saying here that they should specify a file that is _based on_ `ww3_shel.nml` from the WaveWatch III build, but that they have (or their app maintainer has) updated to contain Jinja2 expressions.

Author: It is currently hard-coded within `ww3.py` as such, but perhaps we should explicitly call it `ww3_shel.nml.IN` to clearly differentiate it?

Reviewer: Oh, right -- the _output_ filename is hardcoded, but the input filename is arbitrary. Given that we use the filename `ww3_shel.nml.IN` in our `shared/ww3.yaml`, how about e.g.
```
The path to the input template file containing Jinja2 expressions (perhaps named ``ww3_shel.nml.IN``), based on the ``ww3_shel.nml`` file from the WaveWatchIII build.
```
? It's not IMO really any of our business what they name their input template file, but if this makes the meaning clearer, I'm all for it.

> .. important:: The WaveWatchIII namelist file is provisioned by rendering an input template file containing Jinja2 expressions. Unlike namelist files provisioned by ``uwtools`` for other components, the WaveWatchIII namelist file will not be validated.Note that the non-namelist ``ww3_shel.inp`` file may not be used as the basis for the input template file.:

Reviewer: ```suggestion
.. important:: The WaveWatchIII namelist file is provisioned by rendering an input template file containing Jinja2 expressions. Unlike namelist files provisioned by ``uwtools`` for other components, the WaveWatchIII namelist file will not be validated.
```
I think the note about the bad `ww3_shel.inp` file makes better sense adjacent to the info about the good `ww3_shel.nml`, below.

  **template_values:**

>   The set of key-value pairs that will render the appropriate XML entries in the template input file.

Reviewer: ```suggestion
  Key-value pairs necessary to render all Jinja2 expressions in the input template file named by ``template_file:``.
```
Is there XML in play here? I don't recall seeing any.

Author: cut-and-paste-itis strikes again

>   The path to the input template file containing Jinja2 expressions (perhaps named ``ww3_shel.nml.IN``), based on the ``ww3_shel.nml`` file from the WaveWatchIII build.

Reviewer: ```suggestion
  The path to the input template file containing Jinja2 expressions (perhaps named ``ww3_shel.nml.IN``), based on the ``ww3_shel.nml`` file from the WaveWatchIII build. Note that the non-namelist ``ww3_shel.inp`` file may not be used as the basis for the input template file.
```
